nexus/go-handler.md

```diff
@@ -0,0 +1,307 @@
+	// The workflow ID used here will be returned as the operation ID by default.
+	client.WorkflowStartOptions
+	// Type of workflow to start.
+	WorkflowType string
+	// Input to provide to the workflow.
+	Input        []any
+	// TODO: Should Temporal Header be settable here?
+}
+
+// Returns a Handler that can maps and operation to a workflow that can be started and canceled.
+// Handles HTTP request to payload deserialization for input of type `T`.
+func NewWorkflowRunHandler[T any](func(context.Context, T, *StartOperationRequest) (*WorkflowStartOptions, error)) Handler
+```
+
+**Basic usage:**
+
+```go
+w.RegisterOperation("provision-cell", NewWorkflowRunHandler(
+	func(ctx context.Context, input Input, request *StartOperationRequest) (*WorkflowStartOptions, error) {
+		return &WorkflowStartOptions{
+			Input:        []any{input},
+			WorkflowType: "provision-cell",
+			StartWorkflowOptions: StartWorkflowOptions{
+				ID: fmt.Sprintf("%s-%s", request.Operation, input.CellID),
+			},
+		}, nil
+	}))
+```
+
+### Workflow Query
+
+```go
+package operation // go.temporal.io/sdk/operation
+
+type WorkflowQueryOptions struct {
+	WorkflowID           string
+	// Optionally expose RunID too
+	QueryType            string
+	QueryRejectCondition enums.QueryRejectCondition
+	// Input to provide to the query.
+	Input                []any
+	// TODO: Should Temporal Header be settable here?
+}
+
+func NewWorkflowQueryHandler[T any](func(context.Context, T, *StartOperationRequest) (*WorkflowQueryOptions, error)) Handler
+```
+
+**Basic usage:**
+
+```go
+w.RegisterOperation("get-cell-status", NewWorkflowQueryHandler(
+	func(ctx context.Context, input Input, request *StartOperationRequest) (*WorkflowQueryOptions, error) {
+		return &WorkflowQueryOptions{
+			Input:      []any{input},
+			QueryType:  "get-cell-status",
+			WorkflowID: fmt.Sprintf("%s-%s", "provision-cell", input.CellID),
+		}, nil
+	}))
+```
+
+### Workflow Signal
+
+```go
+package operation // go.temporal.io/sdk/operation
+
+type WorkflowSignalOptions struct {
+	WorkflowID           string
+	// Optionally expose RunID too
+	SignalType            string
+	// Input to provide to the query.
+	Input                []any
+	// TODO: Should Temporal Header be settable here?
+}
+
+func NewWorkflowSignalHandler[T any](func(context.Context, T, *StartOperationRequest) (*WorkflowSignalOptions, error)) Handler
+```
+
+**Basic usage:**
+
+```go
+w.RegisterOperation("update-cell-config", NewWorkflowSignalHandler(
+	func(ctx context.Context, input Input, request *StartOperationRequest) (*WorkflowSignalOptions, error) {
+		return &WorkflowSignalOptions{
+			Input:      []any{input},
+			SignalType: "update-cell-config",
+			WorkflowID: fmt.Sprintf("%s-%s", "provision-cell", input.CellID),
+		}, nil
+	}))
+```
+
+### Workflow Update
+
+Not defined yet. We can easily support sync update, but async update would require more thought on the guarantees
```

Comment:
For sync update, and all the issues about ordering guarantees between signals and updates, and updates from a single source, will nexus affect the ordering? 
Nexus could return a 201 for an operation in which the handler is doing a sync update, making a sync operation async...
I think that eventually nexus will need to serialize updates per workflow if we want to 
use sync updates in a sensible manner

Reply:
Not sure what you mean here, the handler is only responsible for sending the update request, ordering guarantees is out of scope for this proposal. The challenge with async update is ensuring that the callback is delivered even if the update doesn't complete or the workflow is terminated midway.

Reply:
I was thinking ordering end-to-end: a client triggers nexus op1 and gets a 201, then it triggers nexus op2 and gets another 201 (or 200). If the handler is doing updates on the same workflow, people may assume that op1 update happened before op2 update, but is that something we can guarantee with Nexus? Is the handler going to reorder requests?

Reply:
Handler is basically a stateless HTTP request handler, it's not going to be able to do much, and it's not the handler's job.

Reply:
I understand... I'm just thinking how we are going to get the end-to-end ordering back for updates, and it is not going to be easy... 
And if people assume that this is implementing "cross-namespace updates", ordering is the first thing they are going to be surprised about...